### PR TITLE
Avoid incorrectly reusing assertion nodes from property assignments

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/60573 (a regression from https://github.com/microsoft/TypeScript/pull/59282 )

### CODE REVIEW

File: src/compiler/expressionToTypeNode.ts [hunk 1]
```diff
@@ -711,7 +712,10 @@ export function createSyntacticTypeNodeBuilder(
                 undefined;
 
             if (type && !isConstTypeReference(type)) {
-                result = serializeExistingTypeNode(type, context);
+                const assertionNode = serializeExistingTypeNode(type, context);
+                if (resolver.canReuseTypeNodeAnnotation(context, node, getOriginalNode(assertionNode, isTypeNode), symbol)) {
```
Comment (Andarist) [inline src/compiler/expressionToTypeNode.ts:716]: I was somewhat surprised that reusing an existing type node produces its copy with `.original` set on it. But also tbf I don't know when those copies are created in the first place so perhaps that's fine and expected. cc @dragomirtitian

Comment (dragomirtitian) [reply src/compiler/expressionToTypeNode.ts:716]: Copies are always created, this is expected. @weswigham might have better insight into why.

**Comment by jakebailey** [comment]
If this fixes #60573, we should definitely add a dts test that shows that things are fixed.

File: src/compiler/expressionToTypeNode.ts [hunk 2]
```diff
@@ -706,12 +706,12 @@ export function createSyntacticTypeNodeBuilder(
...
             const initializer = node.initializer;
-            const type = isJSDocTypeAssertion(initializer) ? getJSDocTypeAssertionType(initializer) :
+            const assertionNode = isJSDocTypeAssertion(initializer) ? getJSDocTypeAssertionType(initializer) :
                 initializer.kind === SyntaxKind.AsExpression || initializer.kind === SyntaxKind.TypeAssertionExpression ? (initializer as AsExpression | TypeAssertion).type :
                 undefined;
 
-            if (type && !isConstTypeReference(type)) {
-                result = serializeExistingTypeNode(type, context);
+            if (assertionNode && !isConstTypeReference(assertionNode) && resolver.canReuseTypeNodeAnnotation(context, node, assertionNode, symbol)) {
+                result = serializeExistingTypeNode(assertionNode, context);
```
Comment (jakebailey) [inline src/compiler/expressionToTypeNode.ts:714]: I was largely unsure about this given you ask a question about `node` then use that to work with `assertionNode`, but this is effectively the same check as:

```ts
    function typeFromPropertyAssignment(node: PropertyAssignment | ShorthandPropertyAssignment, symbol: Symbol, context: SyntacticTypeNodeBuilderContext) {
        const typeAnnotation = getEffectiveTypeAnnotationNode(node);
        let result;
        if (typeAnnotation && resolver.canReuseTypeNodeAnnotation(context, node, typeAnnotation, symbol)) {
            result = serializeExistingTypeNode(typeAnnotation, context);
        }
```

So I definitely like this change more than the initial version.

**Comment by jakebailey** [comment]
@typescript-bot test it

**Comment by jakebailey** [comment]
@typescript-bot pack this

File: tests/baselines/reference/declarationAssertionNodeNotReusedWhenTypeNotEquivalent1.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+export const unwrapped = unwrap(value);
+
+
+
+
+//// [declarationAssertionNodeNotReusedWhenTypeNotEquivalent1.d.ts]
+export declare const unwrapped: {
+    prop1: "hello";
+};
```
Comment (Andarist) [inline tests/baselines/reference/declarationAssertionNodeNotReusedWhenTypeNotEquivalent1.js:34]: @jakebailey here is the requested dts test

**Review by weswigham — APPROVED** [review]
Yeah, this looks solid now. It'd probably be relevant to ask what output this limits under `isolatedDeclarations`, but... 🤷‍♂️ it kinda is what it is.

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Review by jakebailey — APPROVED** [review]

**Comment by jakebailey** [comment]
@typescript-bot cherry-pick this to release-5.7

**Pull request merged**